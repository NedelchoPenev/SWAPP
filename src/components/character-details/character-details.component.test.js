import React from 'react';
import { shallow, mount } from 'enzyme';

import CharacterDetails from './character-details.component';

describe('CharacterDetails component', () => {
  let wrapper;
  const imageUrl = 'www.testImage.com';
  const mockName = 'Obi-Wan Kenobi';
  const mockHeight = '182';
  const mockMass = '77';
  const mockSpecies = {name: 'Human'};
  const mockHomeworld = {name: 'Stewjon'};
  const mockStarships = [{
    node: {
      id: "starships.48",
      image: 'www.testImage.com',
      name: "Starship",
    }
  }]
  let mockHistory;
  let mockProps

  beforeEach(() => {
    mockHistory = { push: jest.fn() };

    mockProps = {
      person: {
        name: mockName,
        image: imageUrl,
        height: mockHeight,
        mass: mockMass,
        species: mockSpecies,
        homeworld: mockHomeworld,
        starships: { edges: mockStarships }
      },
      history: mockHistory
    };

    wrapper = shallow(<CharacterDetails.WrappedComponent {...mockProps} />);
  });

  it('should render CharacterDetails component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call history.push with the right string when StarshipCard clicked', () => {
    wrapper.find('StarshipCard').simulate('click');

    expect(mockHistory.push).toHaveBeenCalledWith(
      `/starships/${mockStarships[0].node.id}`
    );
  });

  it('should render imageUrl if image is available', () => {
    const newWrapper = mount(<CharacterDetails.WrappedComponent {...mockProps}/>)

    expect(newWrapper.props().person.image).toBe(imageUrl);
  });

  it('should render MISSING_IMG if image is not available', () => {
    mockProps = {
      person: {
        name: mockName,
        image: null,
        height: mockHeight,
        mass: mockMass,
        species: mockSpecies,
        homeworld: mockHomeworld,
        starships: { edges: mockStarships }
      },
      history: mockHistory
    };

    const newWrapper = mount(<CharacterDetails.WrappedComponent {...mockProps}/>)

    expect(newWrapper.props().person.image).toBe(null);
  });

  it('should render height if height prop is available', () => {

    expect(wrapper.find('#pointer').length).toBe(4);

    expect(wrapper.find('#content').first().text()).toEqual(mockHeight);
  });

  it('should NOT render if height is NOT available', () => {
    mockProps = {
      person: {
        name: mockName,
        image: imageUrl,
        height: null,
        mass: mockMass,
        species: mockSpecies,
        homeworld: mockHomeworld,
        starships: { edges: mockStarships }
      },
      history: mockHistory
    };

    const newWrapper = shallow(<CharacterDetails.WrappedComponent {...mockProps}/>)

    expect(newWrapper.find('#pointer').length).toBe(3);

    expect(newWrapper.find('#content').first().text()).not.toEqual(mockHeight);
  });

  it('should render weight if mass prop is available', () => {

    expect(wrapper.find('#pointer').length).toBe(4);

    expect(wrapper.find('#content').at(1).text()).toEqual(mockMass);
  });

  it('should NOT render if weight is NOT available', () => {
    mockProps = {
      person: {
        name: mockName,
        image: imageUrl,
        height: mockHeight,
        mass: null,
        species: mockSpecies,
        homeworld: mockHomeworld,
        starships: { edges: mockStarships }
      },
      history: mockHistory
    };

    const newWrapper = shallow(<CharacterDetails.WrappedComponent {...mockProps}/>)

    expect(newWrapper.find('#pointer').length).toBe(3);

    expect(newWrapper.find('#content').at(1).text()).not.toEqual(mockMass);
  });

  it('should render species if species prop is available', () => {

    expect(wrapper.find('#pointer').length).toBe(4);

    expect(wrapper.find('#content').at(2).text()).toEqual(mockSpecies.name);
  });

  it('should NOT render if species is NOT available', () => {
    mockProps = {
      person: {
        name: mockName,
        image: imageUrl,
        height: mockHeight,
        mass: mockMass,
        species: null,
        homeworld: mockHomeworld,
        starships: { edges: mockStarships }
      },
      history: mockHistory
    };

    const newWrapper = shallow(<CharacterDetails.WrappedComponent {...mockProps}/>)

    expect(newWrapper.find('#pointer').length).toBe(3);

    expect(newWrapper.find('#content').at(2).text()).not.toEqual(mockSpecies.name);
  });

  it('should render homeworld if homeworld prop is available', () => {

    expect(wrapper.find('#pointer').length).toBe(4);

    expect(wrapper.find('#content').at(3).text()).toEqual(mockHomeworld.name);
  });

  it('should NOT render if homeworld is NOT available', () => {
    mockProps = {
      person: {
        name: mockName,
        image: imageUrl,
        height: mockHeight,
        mass: mockMass,
        species: mockSpecies,
        homeworld: null,
        starships: { edges: mockStarships }
      },
      history: mockHistory
    };

    const newWrapper = mount(<CharacterDetails.WrappedComponent {...mockProps}/>)

    expect(newWrapper.find('#pointer').length).toBe(3);
    
    expect(newWrapper.props().person.homeworld).toBe(null);
  });
});