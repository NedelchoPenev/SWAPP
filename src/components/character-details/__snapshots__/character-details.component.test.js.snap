// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetails component should render CharacterDetails component 1`] = `
<styled.div>
  <h1>
    Obi-Wan Kenobi
  </h1>
  <styled.div>
    <styled.div>
      <h2>
        Obi-Wan Kenobi
      </h2>
      <img
        alt="Obi-Wan Kenobi"
        src="www.testImage.com"
      />
      <p>
        <span
          id="pointer"
        >
          Height: 
        </span>
        <span
          id="content"
        >
          182
        </span>
      </p>
      <p>
        <span
          id="pointer"
        >
          Weight: 
        </span>
        <span
          id="content"
        >
          77
        </span>
      </p>
      <p>
        <span
          id="pointer"
        >
          Species: 
        </span>
        <span
          id="content"
        />
      </p>
      <p>
        <span
          id="pointer"
        >
          Home World: 
        </span>
        <span
          id="content"
        />
      </p>
    </styled.div>
    <styled.div>
      <h2>
        Piloted Starships
      </h2>
      <StarshipCard
        key="starships.48"
        onClick={[Function]}
        starship={
          Object {
            "id": "starships.48",
            "image": "www.testImage.com",
            "name": "Starship",
          }
        }
      />
    </styled.div>
  </styled.div>
</styled.div>
`;
